File: scripts/spline/modular/simulate.jl
using DrWatson
@quickactivate :ens

using RCall
using DataFrames
using DataFramesMeta
using Arrow
using TOML

pardir = ARGS[1]
inpath = ARGS[2]
outpath = ARGS[3]
n = TOML.parsefile(pardir * "/params.toml")["mkdf"]["n_sims"]

simulations = rcopy(R"readRDS($inpath)")
data = load_data(:last)


df = DataFrame(index1=Float64[], index2=Float64[], group=String[], reference=String[], landmark=Union{Nothing, String}[], fake=Vector{Vector{Vector{Float64}}}[])

function Base.vec(x::Float64)
	if isempty(x[1])
		return Float64[]
	end
	return Float64[x]
end

function extract(fake)
	result = (collect ∘ values)(fake)
	[vec.(r) for r in result]
end


for row in simulations
	index1 = row[:index1]
	index2 = ismissing(row[:index2]) ? NaN : row[:index2] 
	group = row[:group]
	reference = row[:reference]
	landmark = row[:landmark]
	fake = extract(row[:fake])
	if length(@where(data, :index .== index1).lift[1]) == length(fake) # TODO: make correlation dependent on trials 
		transformed = [[trial[i] for trial in fake] for i in 1:n]
		if !isempty(transformed)
			push!(df, [index1, index2, group, reference, landmark, transformed])
		end
	end
end

df = filter(x->!all(isempty.(x.fake)), df)
Arrow.write(outpath, df)

File: scripts/spline/modular/simulate.R
library(STAR)
library(configr)

sim.all <- function(file, n) {
	res = read_rdata(file)
	lfun = lapply(res$rnfun, mkSelf)[1] 
	trials = lapply(split(res$data, res$data$trial), function(x) x[, names(res$gss$mf)])
	fake = lapply(trials, function(x) replicate(n, sim.catch(res$gss, lfun, x)))
	list(fake=fake, index1=res$index1, index2=res$index2, group=res$group, reference=res$reference, landmark=res$landmark)
}

sim.catch <- function(gss, lfun, x) {
	tryCatch(as.double(thinProcess(object=gss, m2uFctList=lfun, trueData=x, formerSpikes=with(x, time[match(1, event)]))),
			 error = function(e) list())
}

read_rdata <- function(file) {
	r_data <- load(file)
	x <- get(r_data)
	return(x)
}


# inpath = "data/analyses/spline/batch-artificial/best-neigh/out/data/" 

args <- commandArgs(trailingOnly = TRUE)
configpath = paste(args[1], "params.toml", sep="/")
inpath = paste(args[2], "out", "data", sep="/")
outpath = args[3]

config = read.config(file = configpath)
n_sims = config$mkdf$n_sims
infiles <- list.files(path=inpath, pattern=".R", full.names=T, all.files=T)
simulations = lapply(infiles, function(x) sim.all(x, n_sims))


# outpath = "data/analyses/spline/batch-8/best-neigh/post-proc/simulated.rds"
saveRDS(simulations, outpath)
